backend/app/services/attendance_record_service.py

```
@@ -56,113 +59,81 @@ async def clock_in(
+        
         return AttendanceRecordOut.model_validate(db_record)
 
-    except AttendanceError as e:
-        logger.error(f"Attendance error during clock-in for user_id {user.user_id}: {str(e)}", extra={"request_id": request_id})
-        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
-    except ValidationError as e:
-        logger.error(f"Validation error during clock-in for user_id {user.user_id}: {str(e)}", extra={"request_id": request_id})
+    except (AttendanceError, ValidationError) as e:
+        await db.rollback()
+        logger.error(f"Clock-in validation error for user_id {user.user_id}: {str(e)}", 
+                    extra={"request_id": request_id})
```

Ruff (0.12.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)